test/run-spec-wasm2c.py

```diff
@@ -163,6 +183,22 @@ def _CacheModulePrefixes(self):
 
                 self.module_prefix_map[name_idx] = name
 
+    def _WriteModuleImports(self):
+        idx = 0
+        for command in self.commands:
+            if IsModuleCommand(command):
+                name = os.path.basename(command['filename'])
+                module_instance_type = MangleStateInfoTypeName(name)
+                module_prefix = self.module_prefix_map[idx]
+
+                self.out_file.write("static void %s_spectest_print_i32(%s *module_instance, uint32_t i) { spectest_print_i32(i); }\n" % (module_prefix, module_instance_type))
+                self.out_file.write("void (*%sZ_spectestZ_print_i32Z_vi)(%s *, uint32_t) = & %s_spectest_print_i32;\n" % (module_prefix, module_instance_type, module_prefix))
+                self.out_file.write("static void %s_spectest_print(%s *module_instance) { spectest_print(); }\n" % (module_prefix, module_instance_type))
+                self.out_file.write("void (*%sZ_spectestZ_printZ_vv)(%s *) = & %s_spectest_print;\n" % (module_prefix, module_instance_type, module_prefix))
+                self.out_file.write("\n")
+
+                idx += 1
+
```

Comment:
Why are these testing changes necessary? I would hope that the tests would keep running as before, except perhaps with extra code to create an instance, but nothing more.

Reply:
`_WriteModuleImports` converts a function defined in `spec-wasm2c-prefix.c` to its corresponding version that takes a pointer to the module instance as its first argument and is name-mangled by the module name, but does not change how the function executes. Since each module defined in a `wast` file has a different instance type and name for imported functions, we need to repeat `_WriteModuleImports` for each module.

We think functions imported from the host should also take a pointer to the instance as its argument and be name-mangled by module names for the following reasons:
- Hosts should have access to instances
- Hosts should be able to choose which function they want to provide to a certain module

There are potentially other ways of adapting `spec-wasm2c-prefix.c`. What are your thoughts?

Reply:
Thanks - I think I see, but I'm not sure.

Can you maybe paste the output code before and after, or a diff? I think maybe comparing them will be easier to see what you mean here. (In particular, I'm not sure what "Hosts should have access to instances" means, and how it is handled concretely in the code.)

Reply:
The following example is from module 5 of `start.wast`. The following code snippets are taken from `start.5.h`. The module imports `spectest_print_i32`.

**Before the current PR**
```c
/* import: 'spectest' 'print_i32' */
extern void (*Z_spectestZ_print_i32Z_vi)(u32);
#ifdef __cplusplus
```

**After the current PR**
```c
/* import: 'spectest' 'print_i32' */
extern void (*WASM_RT_ADD_PREFIX(Z_spectestZ_print_i32Z_vi))(Z_startZ2E5_module_instance_t *, u32);
#ifdef __cplusplus
```

After the current PR, the imported function is name-mangled using `WASM_RT_ADD_PREFIX` and takes `Z_startZ2E5_module_instance_t *` as its first argument where `Z_startZ2E5_module_instance_t` is the type of the module instance of this module.

---
`spectest_print_i32` is defined in `spectest-wasm2c-prefix.c` as following
```c
static void spectest_print_i32(uint32_t i) {
  printf("spectest.print_i32(%d)\n", i);
}

void (*Z_spectestZ_print_i32Z_vi)(uint32_t) = &spectest_print_i32;
```
Content of `spectest-wasm2c-prefix.c` is copied into the output `main` file of `run-spec-wasm2c.py`.

---
In order to make `run-spec-wasm2c.py` output imported functions that work, we changed `run-spec-wasm2c.py` to instantiate those imported functions. And the output `start-main.c` looks like this:
```c
static void Z_start_5_spectest_print_i32(Z_startZ2E5_module_instance_t *module_instance, uint32_t i) { spectest_print_i32(i); }
void (*Z_start_5Z_spectestZ_print_i32Z_vi)(Z_startZ2E5_module_instance_t *, uint32_t) = & Z_start_5_spectest_print_i32;
```
Essentially, it defines a `Z_start_5_spectest_print_i32` that takes a pointer to the module instance and a `u32`, and calls `spectest_print_i32` on the `u32`. And then, it sets `WASM_RT_ADD_PREFIX(Z_spectestZ_print_i32Z_vi)` to point to `Z_start_5_spectest_print_i32`.

Reply:
I see, thanks!

That all makes sense to me. The name mangling by `WASM_RT_ADD_PREFIX` should have been there before, I think, and adding an instance parameter is clearly necessary.